### ISSUE TITLE
Tap-hold is locked to the hold binding even when tapped until restarted

### ISSUE THREAD

**Issue by zac-ng** [kmonad/kmonad#251]
After running kmonad for a while, occasionally tap-hold will always run the hold binding even if I don't hold the key.  For example, I have a tap-hold on A which runs shift-A if I hold the A button.  If I try pressing A down very quickly it will always print out the capital A instead of the lower case.  This happens to all of my tap-hold bindings, and the only way I can stop it is by restarting kmonad.

**Comment by david-janssen** [comment]
Oh, that is strange.

Are you on the most recent version of master?

Could you run kmonad with `--log-level debug` and copy-paste some of the goings-on when this occurs? I'd like to help, but I need a bit more information. Also, maybe copy-paste your config?

**Comment by zac-ng** [comment]
No, sorry I'm running on the last binary.  I can't seem to find the binary after running stack build.

I'll put my config below.  The next time I run into this problem I'll paste the debug log as well.

(defcfg
        input  (device-file "/dev/input/by-path/pci-0000:2d:00.3-usb-0:3:1.0-event-kbd")
   output (uinput-sink "My KMonad output" "/bin/sleep 2 && /bin/setxkbmap -option compose:ralt")
  
        ;; Comment this is you want unhandled events not to be emitted
        fallthrough true
        ;; Set this to false to disable any command-execution in KMonad
        allow-cmd true
)

(defsrc
  grv  1    2    3    4    5    6    7    8    9    0    -    =    bspc
  tab  q    w    e    r    t    y    u    i    o    p    [    ]    \
  caps a    s    d    f    g    h    j    k    l    ;    '    ret
  lsft z    x    c    v    b    n    m    ,    .    /    rsft
  lctl lmet lalt           spc            ralt rmet cmp  rctl
)

(deflayer qwerty
  @shift`  @shift1    @shift2    @shift3    @shift4    @shift5    @shift6    @shift7    @shift8    @shift9    @shift0    @shift-    @shift=    bspc
  tab  @shiftQ    @shiftW   @shiftE  @shiftR  @shiftT  @shiftY  @shiftU  @shiftI  @shiftO  @shiftP   @shift[    @shift]    @shift\
  rctl @shiftA   @shiftS  @shiftD  @shiftF  @shiftG  @shiftH  @shiftJ  @shiftK  @shiftL   @shift;    @shift'    ret
  lsft @shiftZ  @shiftX  @shiftC  @shiftV  @shiftB  @shiftN  @shiftM   @shift,    @shift.    @shift/    rsft
  lctl lmet lalt           spc            ralt rmet cmp caps
)

(defalias
        shiftA (tap-hold 200 a A)
        shiftB (tap-hold 200 b B)
        shiftC (tap-hold 200 c C)
        shiftD (tap-hold 200 d D)
        shiftE (tap-hold 200 e E)
        shiftF (tap-hold 200 f F)
        shiftG (tap-hold 200 g G)
        shiftH (tap-hold 200 h H)
        shiftI (tap-hold 200 i I)
        shiftJ (tap-hold 200 j J)
        shiftK (tap-hold 200 k K)
        shiftL (tap-hold 200 l L)
        shiftM (tap-hold 200 m M)
        shiftN (tap-hold 200 n N)
        shiftO (tap-hold 200 o O)
        shiftP (tap-hold 200 p P)
        shiftQ (tap-hold 200 q Q)
        shiftR (tap-hold 200 r R)
        shiftS (tap-hold 200 s S)
        shiftT (tap-hold 200 t T)
        shiftU (tap-hold 200 u U)
        shiftV (tap-hold 200 v V)
        shiftW (tap-hold 200 w W)
        shiftX (tap-hold 200 x X)
        shiftY (tap-hold 200 y Y)
        shiftZ (tap-hold 200 z Z)
        shift` (tap-hold 200 ` ~)
        shift1 (tap-hold 200 1 !)
        shift2 (tap-hold 200 2 @)
        shift3 (tap-hold 200 3 #)
        shift4 (tap-hold 200 4 $)
        shift5 (tap-hold 200 5 %)
        shift6 (tap-hold 200 6 ^)
        shift7 (tap-hold 200 7 &)
        shift8 (tap-hold 200 8 *)
        shift9 (tap-hold 200 9 \()
        shift0 (tap-hold 200 0 \))
        shift- (tap-hold 200 - \_)
        shift= (tap-hold 200 = +)
        shift[ (tap-hold 200 [ {)
        shift] (tap-hold 200 ] })
        shift\ (tap-hold 200 \ |)
        shift; (tap-hold 200 ; :)
        shift' (tap-hold 200 ' ")
        shift, (tap-hold 200 , <)
        shift. (tap-hold 200 . >)
        shift/ (tap-hold 200 / ?)
)

**Comment by david-janssen** [comment]
That is certainly a very peculiar config you have, not many people run things like that, I think, so you might very well be the first person to have walked into a strange bug. I'm looking forward to the debug output.

> No, sorry I'm running on the last binary. I can't seem to find the binary after running stack build.

When you `stack build` the binary gets put in a local `.stack-work` directory. If you `stack install` stack then copies that binary to a logical local bin directory. I think by default it goes to `~/.local/bin`.

**Comment by 6gk** [comment]
Hey, @zac-ng just so you know, you can put \`\`\` above & below parts of comments (on their own line) to make the font monospace, it makes reading configs a bit easier and prevents accidentally pinging people.

![example](https://user-images.githubusercontent.com/40295114/118594329-9d0d6e00-b7ec-11eb-92c0-280d763ce917.png)

**Comment by david-janssen** [comment]
Oh yeah, just noticed, you must've pinged a *bunch* of people... Hahaha :)

**Comment by zac-ng** [comment]
Oops, yeah didn't notice that.  I'm running the latest version of master now, I'll post my log when I run into the error again.

**Comment by zac-ng** [comment]
Here is some of what my debug log is printing out when I ran into the error.
```
--------------------------------------------------------------------------------
Received event: Release <e>
Running hooks
Registering untimed hook: 9217
Registering untimed hook: 9218
Emitting: Press <e>
Unblocking input stream, no stored events
Emitting: Release <e>
Tried cancelling expired hook: 9215

--------------------------------------------------------------------------------
Received event: Press <r>
Running hooks
Registering untimed hook: 9219
Registering untimed hook: 9220
Block level set to: 1
Registering untimed hook: 9222
Registering 200ms hook: 9221

--------------------------------------------------------------------------------
Received event: Release <r>
Running hooks
Registering untimed hook: 9223
Registering untimed hook: 9224
Emitting: Press <r>
Unblocking input stream, no stored events
Emitting: Release <r>
Tried cancelling expired hook: 9221

--------------------------------------------------------------------------------
Received event: Press <j>
Running hooks
Registering untimed hook: 9225
Registering untimed hook: 9226
Block level set to: 1
Registering untimed hook: 9228
Registering 200ms hook: 9227

--------------------------------------------------------------------------------
Received event: Release <j>
Running hooks
Registering untimed hook: 9229
Registering untimed hook: 9230
Emitting: Press <j>
Unblocking input stream, no stored events
Emitting: Release <j>
Tried cancelling expired hook: 9227

--------------------------------------------------------------------------------
Received event: Press <d>
Running hooks
Registering untimed hook: 9231
Registering untimed hook: 9232
Block level set to: 1
Registering untimed hook: 9234
Registering 200ms hook: 9233

--------------------------------------------------------------------------------
Received event: Release <d>
Running hooks
Registering untimed hook: 9235
Emitting: Press <d>
Unblocking input stream, no stored events
Emitting: Release <d>
Registering untimed hook: 9236
Tried cancelling expired hook: 9233

--------------------------------------------------------------------------------
Received event: Press <k>
Running hooks
Registering untimed hook: 9237
Registering untimed hook: 9238
Block level set to: 1
Registering untimed hook: 9240
Registering 200ms hook: 9239

--------------------------------------------------------------------------------
Received event: Release <k>
Running hooks
Registering untimed hook: 9241
Registering untimed hook: 9242
Emitting: Press <k>
Unblocking input stream, no stored events
Emitting: Release <k>
Tried cancelling expired hook: 9239

--------------------------------------------------------------------------------
Received event: Press <i>
Running hooks
Registering untimed hook: 9243
Registering untimed hook: 9244
Block level set to: 1
Registering untimed hook: 9246
Registering 200ms hook: 9245

--------------------------------------------------------------------------------
Received event: Release <i>
Running hooks
Registering untimed hook: 9247
Registering untimed hook: 9248
Emitting: Press <i>
Unblocking input stream, no stored events
Emitting: Release <i>
Tried cancelling expired hook: 9245


```

**Comment by david-janssen** [comment]
Thanks for the log. I'm afraid that looks really normal, which of course is annoying, since your problem is real. So, diagnosis-wise we've excluded 'all buttons get into a weird state' or 'kmonad has a too-many-hooks' problem or something.

There is another possible explanation for only getting capitals: at some point somewhere a shift was pressed, and never released. We're going to need some more data :-). Could you, the next time it goes wrong, look up into the log and find the last time KMonad emitted a shift?

Maybe the easiest way to do this would be to `> logfile.txt` the output (better support for logging is coming, but isn't there yet). Then, when the issue happens again, close kmonad and open your logfile in a text-editor. You should be able to find the last time where KMonad emitted either a left or right shift.

Could you then cut and paste that bit with some stuff from before and after that?

Don't upload the file though! And make sure to delete it, because you are now essentially running a keylogger, and this could reveal any passwords you might've typed while running KMonad.

**Comment by MaxGyver83** [comment]
Only after reading David's comment, I recognize that I probably have the same issue. In my case, this (=only uppercase letters) happens after I have left my laptop, the screen has turned black and I wake it up again. It might be related that I usually wake it up by pressing a shift key. I don't know if bluetooth falls asleep, too. Then, a udev rule runs a script that restarts KMonad. Next time it happens, I'll check the KMonad log file and if udev was triggered again.

**Comment by david-janssen** [comment]
>  It might be related that I usually wake it up by pressing a shift key. I don't know if bluetooth falls asleep, too. Then, a udev rule runs a script that restarts KMonad. Next time it happens, I'll check the KMonad log file and if udev was triggered again.

Oh yeah, that sounds *really* plausible. Maybe I can figure out how to detect 'going to suspend' from KMonad and act accordingly, although that would be a fix that I would only investigate after the new keycode-refactor branch has been tested cross platform and merged into master.

**Comment by zac-ng** [comment]
Yeah, I'll copy and paste from the log file when I run into the error again.

Just to note this happens to me with the symbols as well (as I have a tap-hold rule for them as well).  I have a wired keyboard for reference, and this happens when I've been working a while.  It never happens when I stops suspend, it only occurs in the middle of working.

**Comment by zac-ng** [comment]
How can I tell where the shift occurs?  lshft and rsft doesn't show up in the log file.

**Comment by david-janssen** [comment]
> How can I tell where the shift occurs? lshft and rsft doesn't show up in the log file.

```
(defsrc
  grv  1    2    3    4    5    6    7    8    9    0    -    =    bspc
  tab  q    w    e    r    t    y    u    i    o    p    [    ]    \
  caps a    s    d    f    g    h    j    k    l    ;    '    ret
  lsft z    x    c    v    b    n    m    ,    .    /    rsft
  lctl lmet lalt           spc            ralt rmet cmp  rctl
)```

They don't show up in the log? But looking at the source you are clearly capturing `lsft` and `rsft`. Then looking at the layer, you are also clearly:
- binding `lsft` and `rsft` to the `lsft` and `srft` source buttons (i.e., in the location in your `deflayer`, those are also the buttons that you've specified)
- you are using a lot of buttons like `A`, which is just an indicator to KMonad to press `press-shift, tap a, release-shift`.

If you aren't getting any mention at all of KMonad emitting shifts when you hold one of your buttons, or when you press the actual shift buttons on your keyboord, then something very strange is going on.

Another way to test if its a single shift being pressed and never released, vs. all the butttons have gone weird is to either:
- Bind 1 button to emit something weird on hold, like `(tap-hold 200 a B)`, then if your problem occurs and that button is giving you `A`s, then it's a shift issue, and when `B`s, it's a tap-hold issue. You'd have to find a button you could rebind easily without it getting in the way of actually using your keyboard well though :)
- If you have another keyboard, you could just tap some buttons on that. If those are *also* capitalized, we know it's a `shift` issue.

**Comment by zac-ng** [comment]
Nope they don't. I actually never press lsft or rsft, any time I need to press shift I just tap-hold the key.  The shift isn't ever emitted in my log (although if I explicitly press shift kmonad still recognizes it).  Here's an example of where I pressed '2', and the '@' was emitted instead.  
```
--------------------------------------------------------------------------------
Received event: Press <2>
Running hooks
Registering untimed hook: 13942
Registering untimed hook: 13943
Block level set to: 1
Registering untimed hook: 13945
Registering 200ms hook: 13944

--------------------------------------------------------------------------------
Received event: Release <2>
Running hooks
Registering untimed hook: 13946
Registering untimed hook: 13947
Emitting: Press <2>
Unblocking input stream, no stored events
Emitting: Release <2>
Tried cancelling expired hook: 13938
Tried cancelling expired hook: 13944

--------------------------------------------------------------------------------
```
There's no mention of shift even though '@' is printed.  I'll try the alternate keyboard as well.

**Comment by david-janssen** [comment]
Then there must be an issue with my logging. Because the way we send `@`s to the system is by this `shift -> 2 -> shift` approach I mentioned. I'll have to investigate this further. Sorry we weren't able to find a straightforward fix quickly.
